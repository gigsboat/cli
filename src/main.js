import path from 'path'
import marked from 'marked'

import { getAllFiles } from './utils/fs.js'
import { convertToJson } from './utils/yaml-parser.js'
import {
  getEventsMd,
  formatToMarkdown,
  getStatsBadges
} from './utils/md-formatter.js'
import { createYearBuckets, getEventsStats } from './utils/content-manager.js'

export { formatToMarkdown, generateGigs, generateDocument }

async function generateDocument({ sourceDirectory, preContent, postContent }) {
  let markdownOutputPreContent = ''
  let markdownOutputPostContent = ''
  let document = ''

  const { entriesForYearMarkdown, entriesByBucket } = await generateGigs({
    sourceDirectory
  })

  if (preContent) {
    markdownOutputPreContent = processCustomContent({
      contents: preContent
    })
  }

  if (postContent) {
    markdownOutputPostContent = processCustomContent({
      contents: postContent
    })
  }

  const statsHeader = `<div align='center'>${marked.parse(
    getStatsBadges(entriesByBucket)
  )}</div>
  `

  const currentDate = new Date().toISOString()
  const footer = `<i>Updated on ${currentDate}</i>`

  document +=
    statsHeader +
    markdownOutputPreContent +
    '\n' +
    entriesForYearMarkdown +
    '\n' +
    markdownOutputPostContent +
    '\n' +
    footer

  return document
}

function processCustomContent({ contents }) {
  let markdownContent = ''
  for (const content of contents) {
    if (content.hasOwnProperty('raw')) {
      markdownContent += content.raw + '\n'
    }

    if (content.hasOwnProperty('format')) {
      markdownContent += formatToMarkdown(content.format) + '\n'
    }
  }

  return markdownContent
}

async function generateGigs({ sourceDirectory }) {
  let directoryPath = path.join(process.cwd(), sourceDirectory)
  if (path.isAbsolute(sourceDirectory)) {
    directoryPath = sourceDirectory
  }

  const allFilesFlatList = await getAllFiles(directoryPath)

  const entries = []
  for (const filePath of allFilesFlatList) {
    const json = await convertToJson(filePath)

    // push the relative path to the file as part of the data object
    let fileDirectoryPath = path.relative(directoryPath, filePath)
    if (!path.isAbsolute(sourceDirectory)) {
      fileDirectoryPath = path.join(sourceDirectory, fileDirectoryPath)
    }
    json.fileRelativePath = fileDirectoryPath

    entries.push(json)
  }

  const entriesByBucket = await getEntriesByBuckets(entries)
  const entriesForYearMarkdown = getEventsMd(entriesByBucket.bucketsByYear)
  return {
    entriesForYearMarkdown,
    entriesByBucket
  }
}

async function getEntriesByBuckets(entries) {
  // @TODO these buckets:
  //  const bucketsLanguage = {}
  //  const bucketsCountry = {}

  const bucketsByYear = createYearBuckets(entries)
  const statsTotal = getEventsStats(entries)

  return {
    bucketsByYear,
    stats: statsTotal
  }
}
