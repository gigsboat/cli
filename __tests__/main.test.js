import { URL } from 'url'
import { jest } from '@jest/globals'
import path from 'path'
import { generateDocument } from '../src/main.js'

const __dirname = new URL('.', import.meta.url).pathname

beforeEach(() => {
  jest.useFakeTimers('modern')
  jest.setSystemTime(new Date(Date.UTC(2021, 11, 14)))
})

afterAll(() => jest.useRealTimers())

test('given a directory of markdown files a full document is rendered', async () => {
  const filePath = path.join(__dirname, './__fixtures__/main-datafiles')
  const jsonResult = await generateDocument({ sourceDirectory: filePath })
  expect(jsonResult).toMatchSnapshot()
})

test('given a directory of markdown files a full document is rendered along with pre and post content', async () => {
  const filePath = path.join(__dirname, './__fixtures__/main-datafiles')
  console.log('---------------------------->')
  console.log(filePath)
  console.log('---------------------------->')
  const jsonResult = await generateDocument({
    sourceDirectory: filePath,
    preContent: [
      {
        raw: '<p> actual html is allowed too </p>'
      },
      {
        format: [
          {
            p: 'Liran Tal'
          }
        ]
      }
    ],
    postContent: [
      {
        raw: '<p align="center"> rendered in postcontent </p>'
      },
      {
        format: [
          {
            p: 'Thank you!'
          }
        ]
      }
    ]
  })
  expect(jsonResult).toMatchSnapshot()
})
