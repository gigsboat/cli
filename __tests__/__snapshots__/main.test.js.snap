// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a directory of markdown files a full document is rendered 1`] = `
"<div align='center'><p><img src=\\"https://img.shields.io/badge/total-6-blue?style=flat-square\\" alt=\\"Total Events\\"> <img src=\\"https://img.shields.io/badge/meetups-1-violet?style=flat-square\\" alt=\\"Total Meetups\\"> <img src=\\"https://img.shields.io/badge/conferences-1-red?style=flat-square\\" alt=\\"Total Conferences\\"> <img src=\\"https://img.shields.io/badge/podcasts-1-yellow?style=flat-square\\" alt=\\"Total Podcasts\\"> <img src=\\"https://img.shields.io/badge/webinars-1-lightgrey?style=flat-square\\" alt=\\"Total Webinars\\"> <img src=\\"https://img.shields.io/badge/articles-1-green?style=flat-square\\" alt=\\"Total Podcasts\\"></p>
</div>
  
# Table of Contents


 - [Year of 2021](#2021) - total events 3
 - [Year of 2019](#2019) - total events 3

# 2021


![Total Events](https://img.shields.io/badge/total-3-blue?style=flat-square)  ![Total Conferences](https://img.shields.io/badge/conferences-1-red?style=flat-square)   ![Total Podcasts](https://img.shields.io/badge/articles-1-green?style=flat-square)


| Date | Event | Title | Slides | Recording | Location | Language |
| ---- | ----- | ----- | ------ | --------- | -------- | -------- |
| 2021-7-21 | Cyber Week | Are We Forever Doomed By Software Supply Chain Risks? |  | [Recording](https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk) | IL | English |
| 2021-6-13 | Cyber Week | Are We Forever Doomed By Software Supply Chain Risks? |  | [Recording](https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk) | IL | English |
| 2021-1-1 | Cyber Week | Are We Forever Doomed By Software Supply Chain Risks? |  | [Recording](https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk) | IL | English |


# 2019


![Total Events](https://img.shields.io/badge/total-3-blue?style=flat-square) ![Total Meetups](https://img.shields.io/badge/meetups-1-violet?style=flat-square)  ![Total Podcasts](https://img.shields.io/badge/podcasts-1-yellow?style=flat-square) ![Total Webinars](https://img.shields.io/badge/webinars-1-lightgrey?style=flat-square) 


| Date | Event | Title | Slides | Recording | Location | Language |
| ---- | ----- | ----- | ------ | --------- | -------- | -------- |
| 2019-9-27 | JSConf Budapest | StrangerDanger: Finding Security Vulnerabilities Before They Find You! |  | [Recording](https://www.youtube.com/watch?v=3H8pF6yoSgU&list=PL37ZVnwpeshEMCvdYDdZ09Sy-toTftWu0&index=15&ab_channel=JSConf) | HU | English |
| 2019-9-15 | JSConf Budapest | StrangerDanger: Finding Security Vulnerabilities Before They Find You! |  | [Recording](https://www.youtube.com/watch?v=3H8pF6yoSgU&list=PL37ZVnwpeshEMCvdYDdZ09Sy-toTftWu0&index=15&ab_channel=JSConf) | HU | English |
| 2019-9-12 | APIdays Barcelona | Consumer-Driven Contracts: A better approach for API Testing | [Slides](https://slides.com/lirantal/consumer-driven-contracts) | [Recording](https://www.youtube.com/watch?v=zfGKX5iKSis&list=PLmEaqnTJ40Oqo9VlbcUakVmFZgx5weLrs&index=25&t=141s&ab_channel=apidays) | ES | English |



"
`;

exports[`given a directory of markdown files a full document is rendered along with pre and post content 1`] = `
"<div align='center'><p><img src=\\"https://img.shields.io/badge/total-6-blue?style=flat-square\\" alt=\\"Total Events\\"> <img src=\\"https://img.shields.io/badge/meetups-1-violet?style=flat-square\\" alt=\\"Total Meetups\\"> <img src=\\"https://img.shields.io/badge/conferences-1-red?style=flat-square\\" alt=\\"Total Conferences\\"> <img src=\\"https://img.shields.io/badge/podcasts-1-yellow?style=flat-square\\" alt=\\"Total Podcasts\\"> <img src=\\"https://img.shields.io/badge/webinars-1-lightgrey?style=flat-square\\" alt=\\"Total Webinars\\"> <img src=\\"https://img.shields.io/badge/articles-1-green?style=flat-square\\" alt=\\"Total Podcasts\\"></p>
</div>
  <p> actual html is allowed too </p>

Liran Tal


# Table of Contents


 - [Year of 2021](#2021) - total events 3
 - [Year of 2019](#2019) - total events 3

# 2021


![Total Events](https://img.shields.io/badge/total-3-blue?style=flat-square)  ![Total Conferences](https://img.shields.io/badge/conferences-1-red?style=flat-square)   ![Total Podcasts](https://img.shields.io/badge/articles-1-green?style=flat-square)


| Date | Event | Title | Slides | Recording | Location | Language |
| ---- | ----- | ----- | ------ | --------- | -------- | -------- |
| 2021-7-21 | Cyber Week | Are We Forever Doomed By Software Supply Chain Risks? |  | [Recording](https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk) | IL | English |
| 2021-6-13 | Cyber Week | Are We Forever Doomed By Software Supply Chain Risks? |  | [Recording](https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk) | IL | English |
| 2021-1-1 | Cyber Week | Are We Forever Doomed By Software Supply Chain Risks? |  | [Recording](https://www.youtube.com/watch?v=x74sMCaZKbg&ab_channel=Snyk) | IL | English |


# 2019


![Total Events](https://img.shields.io/badge/total-3-blue?style=flat-square) ![Total Meetups](https://img.shields.io/badge/meetups-1-violet?style=flat-square)  ![Total Podcasts](https://img.shields.io/badge/podcasts-1-yellow?style=flat-square) ![Total Webinars](https://img.shields.io/badge/webinars-1-lightgrey?style=flat-square) 


| Date | Event | Title | Slides | Recording | Location | Language |
| ---- | ----- | ----- | ------ | --------- | -------- | -------- |
| 2019-9-27 | JSConf Budapest | StrangerDanger: Finding Security Vulnerabilities Before They Find You! |  | [Recording](https://www.youtube.com/watch?v=3H8pF6yoSgU&list=PL37ZVnwpeshEMCvdYDdZ09Sy-toTftWu0&index=15&ab_channel=JSConf) | HU | English |
| 2019-9-15 | JSConf Budapest | StrangerDanger: Finding Security Vulnerabilities Before They Find You! |  | [Recording](https://www.youtube.com/watch?v=3H8pF6yoSgU&list=PL37ZVnwpeshEMCvdYDdZ09Sy-toTftWu0&index=15&ab_channel=JSConf) | HU | English |
| 2019-9-12 | APIdays Barcelona | Consumer-Driven Contracts: A better approach for API Testing | [Slides](https://slides.com/lirantal/consumer-driven-contracts) | [Recording](https://www.youtube.com/watch?v=zfGKX5iKSis&list=PLmEaqnTJ40Oqo9VlbcUakVmFZgx5weLrs&index=25&t=141s&ab_channel=apidays) | ES | English |



<p align=\\"center\\"> rendered in postcontent </p>

Thank you!

"
`;
